src/api/block_builder/input_selection/utxo_chains.rs

```diff
@@ -0,0 +1,218 @@
+                    if let Ok(output_id) = client.foundry_output_id(foundry_output.id()).await {
+                        let output_response = client.get_output(&output_id).await?;
+                        if let OutputDto::Foundry(foundry_output) = &output_response.output {
+                            for unlock_condition in &foundry_output.unlock_conditions {
+                                if let UnlockConditionDto::ImmutableAliasAddress(
+                                    immutable_alias_address_unlock_condition_dto,
+                                ) = unlock_condition
+                                {
+                                    let address =
+                                        Address::try_from(&immutable_alias_address_unlock_condition_dto.address)?;
+                                    utxo_chains.push((address, output_response.clone()));
+                                }
+                            }
+                        }
+                    }
+                }
+                _ => {}
+            }
+        }
+
+        // Get recursively owned alias or nft outputs
+        get_alias_and_nfts_recursively(self.client, &mut utxo_chains).await?;
+
+        let mut utxo_chain_inputs = Vec::new();
+        for (unlock_address, output_response) in utxo_chains {
+            let (address_index, internal) = match self.secret_manager {
+                Some(secret_manager) => {
+                    match unlock_address {
+                        Address::Ed25519(_) => {
+                            search_address(
+                                secret_manager,
+                                &bech32_hrp,
+                                self.coin_type,
+                                self.account_index,
+                                self.input_range.clone(),
+                                &unlock_address,
+                            )
+                            .await?
+                        }
+                        // Alias and NFT addresses can't be generated from a private key
+                        _ => (0, false),
+                    }
+                }
+                None => (0, false),
+            };
+
+            utxo_chain_inputs.push(InputSigningData {
+                output: Output::try_from(&output_response.output)?,
+                output_metadata: OutputMetadata::try_from(&output_response.metadata)?,
+                chain: Some(Chain::from_u32_hardened(vec![
+                    HD_WALLET_TYPE,
+                    self.coin_type,
+                    self.account_index,
+                    internal as u32,
+                    address_index,
+                ])),
+                bech32_address: unlock_address.to_bech32(&bech32_hrp),
+            });
+        }
+
+        Ok(utxo_chain_inputs)
+    }
+}
+
+/// Get recursively owned alias and nft outputs and add them to the utxo_chains
+pub(crate) async fn get_alias_and_nfts_recursively(
```

Comment:
Should be either `get_aliases_and_nfts_recursively` or `get_alias_and_nft_outputs_recursively `